File: static/solutions/java/src/main/java/com/epi/RookAttack.java
package com.epi;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

public class RookAttack {
  // @include
  public static void rookAttack(List<List<Integer>> A) {
    int m = A.size(), n = A.get(0).size();
    boolean hasFirstRowZero = false;
    for (int j = 0; j < n; ++j) {
      if (A.get(0).get(j) == 0) {
        hasFirstRowZero = true;
        break;
      }
    }
    boolean hasFirstColumnZero = false;
    for (int i = 0; i < m; ++i) {
      if (A.get(i).get(0) == 0) {
        hasFirstColumnZero = true;
        break;
      }
    }

    for (int i = 1; i < m; ++i) {
      for (int j = 1; j < n; ++j) {
        if (A.get(i).get(j) == 0) {
          A.get(i).set(0, 0);
          A.get(0).set(j, 0);
        }
      }
    }

    for (int i = 1; i < m; ++i) {
      if (A.get(i).get(0) == 0) {
        Collections.fill(A.get(i), 0);
      }
    }

    for (int j = 1; j < n; ++j) {
      if (A.get(0).get(j) == 0) {
        for (int i = 1; i < m; ++i) {
          A.get(i).set(j, 0);
        }
      }
    }

    if (hasFirstRowZero) {
      Collections.fill(A.get(0), 0);
    }
    if (hasFirstColumnZero) {
      for (int i = 0; i < m; ++i) {
        A.get(i).set(0, 0);
      }
    }
  }
  // @exclude

  private static void checkAns(List<List<Integer>> A, List<List<Integer>> ans) {
    for (int i = 0; i < A.size(); ++i) {
      for (int j = 0; j < A.get(i).size(); ++j) {
        if (A.get(i).get(j) == 0) {
          for (List<Integer> an : ans) {
            assert(an.get(j) == 0);
          }
          for (int k = 0; k < ans.get(i).size(); ++k) {
            assert(ans.get(i).get(k) == 0);
          }
        }
      }
    }
  }

  public static void main(String[] args) {
    Random r = new Random();
    for (int times = 0; times < 1000; ++times) {
      int m, n;
      if (args.length == 2) {
        m = Integer.parseInt(args[0]);
        n = Integer.parseInt(args[1]);
      } else {
        m = r.nextInt(50) + 1;
        n = r.nextInt(50) + 1;
      }
      List<List<Integer>> A = new ArrayList<>(m);
      for (int i = 0; i < m; ++i) {
        A.add(new ArrayList(n));
        for (int j = 0; j < n; ++j) {
          A.get(i).add(r.nextInt(2));
        }
      }
      List<List<Integer>> copyA = new ArrayList<>(m);
      for (List<Integer> a : A) {
        copyA.add(new ArrayList(a));
      }
      System.out.println("m = " + m + ", n = " + n);
      rookAttack(A);
      checkAns(copyA, A);
    }
  }
}

File: solutions/python/rook_attack.py
import sys
import random
import copy


# @include
def rook_attack(A):
    m, n = len(A), len(A[0])
    has_first_row_zero = any(not A[0][j] for j in range(n))
    has_first_column_zero = any(not A[i][0] for i in range(m))

    for i in range(1, m):
        for j in range(1, n):
            if not A[i][j]:
                A[i][0] = A[0][j] = 0

    for i in range(1, m):
        if not A[i][0]:
            for j in range(1, n):
                A[i][j] = 0

    for j in range(1, n):
        if not A[0][j]:
            for i in range(1, m):
                A[i][j] = 0

    if has_first_row_zero:
        for j in range(n):
            A[0][j] = 0

    if has_first_column_zero:
        for i in range(m):
            A[i][0] = 0
# @exclude


def check_ans(A, ans):
    assert all(not ans[k][j] for i in range(len(A)) for j in range(len(A[i]))
               if not A[i][j] for k in range(len(ans)))
    assert all(not ans[i][k] for i in range(len(A)) for j in range(len(A[i]))
               if not A[i][j] for k in range(len(ans[i])))


def main():
    for _ in range(1000):
        if len(sys.argv) == 3:
            m = int(sys.argv[1])
            n = int(sys.argv[2])
        else:
            m = random.randint(1, 50)
            n = random.randint(1, 50)
        A = [[random.randint(0, 1) for _ in range(n)] for _ in range(m)]
        copy_A = copy.deepcopy(A)
        print('m = %d, n = %d' % (m, n))
        rook_attack(A)
        check_ans(copy_A, A)


if __name__ == '__main__':
    main()
